datafusion/spark/src/function/hash/utils.rs

```diff
@@ -0,0 +1,614 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::array::{
+    Array, ArrayAccessor, ArrayIter, ArrayRef, ArrowPrimitiveType, BinaryViewArray,
+    BooleanArray, Date32Array, Date64Array, Decimal128Array, Decimal256Array,
+    DictionaryArray, DurationMicrosecondArray, DurationMillisecondArray,
+    DurationNanosecondArray, DurationSecondArray, FixedSizeBinaryArray, Float32Array,
+    Float64Array, GenericBinaryArray, Int16Array, Int32Array, Int64Array, Int8Array,
+    IntervalDayTimeArray, IntervalMonthDayNanoArray, IntervalYearMonthArray,
+    LargeListArray, LargeListViewArray, LargeStringArray, ListArray, ListViewArray,
+    MapArray, PrimitiveArray, StringArray, StringViewArray, StructArray,
+    Time32MillisecondArray, Time32SecondArray, Time64MicrosecondArray,
+    Time64NanosecondArray, TimestampMicrosecondArray, TimestampMillisecondArray,
+    TimestampNanosecondArray, TimestampSecondArray, UInt16Array, UInt32Array,
+    UInt64Array, UInt8Array, UnionArray,
+};
+use arrow::datatypes::{
+    i256, ArrowDictionaryKeyType, DataType, Int16Type, Int32Type, Int64Type, Int8Type,
+    IntervalDayTime, IntervalMonthDayNano, IntervalUnit, TimeUnit, UInt16Type,
+    UInt32Type, UInt64Type, UInt8Type,
+};
+use bigdecimal::num_traits::{Float, ToBytes};
+use datafusion_common::{exec_err, DataFusionError, Result};
+use std::sync::Arc;
+
+fn hash_impl<T, H: Copy>(
+    arr: ArrayIter<impl ArrayAccessor<Item = T>>,
+    seed: &mut [H],
+    func: impl Fn(H, T) -> H,
+) -> Result<()> {
+    let len = arr.len();
+    if len != seed.len() {
+        return exec_err!("Array length mismatch: {} != {}", len, seed.len());
+    }
+    for (hash, elem) in seed.iter_mut().zip(arr) {
+        if let Some(elem) = elem {
+            *hash = func(*hash, elem);
+        }
+    }
+    Ok(())
+}
+
+fn try_hash_impl<T, H: Copy>(
+    arr: ArrayIter<impl ArrayAccessor<Item = T>>,
+    seed: &mut [H],
+    func: impl Fn(H, T) -> Result<H>,
+) -> Result<()> {
+    let len = arr.len();
+    if len != seed.len() {
+        return exec_err!("Array length mismatch: {} != {}", len, seed.len());
+    }
+    for (hash, elem) in seed.iter_mut().zip(arr) {
+        if let Some(elem) = elem {
+            *hash = func(*hash, elem)?;
+        }
+    }
+    Ok(())
+}
+
+pub trait SparkHasher<H: Copy + std::fmt::Debug> {
+    fn oneshot(seed: H, data: &[u8]) -> H;
+
+    fn hash_boolean(arr: &BooleanArray, seed: &mut [H]) -> Result<()> {
+        let true_bytes = 1_i32.to_le_bytes();
+        let false_bytes = 0_i32.to_le_bytes();
+        hash_impl(arr.iter(), seed, |seed, v| {
+            if v {
+                Self::oneshot(seed, &true_bytes)
+            } else {
+                Self::oneshot(seed, &false_bytes)
+            }
+        })?;
+        Ok(())
+    }
+
+    fn hash_primitive<T, U>(arr: &PrimitiveArray<T>, seed: &mut [H]) -> Result<()>
```

Comment:
I would love it if the generic names were more descriptive; seeing `H`, `T` and `U` all over the place but having to remember which does what (especially as `SparkHasher` has a `H` which `hash_impl` has as well, but functions inside `SparkHasher` have `T` and `U` but only `T` is shared with `hash_impl`) gets confusing real fast, especially as there is a lack of any explaining documentation.